internal/blueprint/fsnode_customizations.go

```diff
@@ -0,0 +1,468 @@
+package blueprint
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
+
+	"github.com/osbuild/osbuild-composer/internal/common"
+	"github.com/osbuild/osbuild-composer/internal/fsnode"
+	"github.com/osbuild/osbuild-composer/internal/pathpolicy"
+)
+
+// validateModeString checks that the given string is a valid mode octal number
+func validateModeString(mode string) error {
+	// Check that the mode string matches the octal format regular expression.
+	// The leading is optional.
+	if regexp.MustCompile(`^[0]{0,1}[0-7]{3}$`).MatchString(mode) {
+		return nil
+	}
+	return fmt.Errorf("invalid mode %s: must be an octal number", mode)
+}
+
+// DirectoryCustomization represents a directory to be created in the image
+type DirectoryCustomization struct {
+	// Absolute path to the directory
+	Path string `json:"path" toml:"path"`
+	// Owner of the directory specified as a string (user name), int64 (UID) or nil
+	User interface{} `json:"user,omitempty" toml:"user,omitempty"`
+	// Owner of the directory specified as a string (group name), int64 (UID) or nil
+	Group interface{} `json:"group,omitempty" toml:"group,omitempty"`
+	// Permissions of the directory specified as an octal number
+	Mode string `json:"mode,omitempty" toml:"mode,omitempty"`
```

Comment:
The `User` and `Group` are optional, but not the mode?

Reply:
Not sure what do you mean. `mode` is optional. The default value for a string (empty string) is considered to be an unset value. I could have used pointer to string, but there would be no real benefit of it, and at the same time the use would get harder when using string literals (e.g. in unit tests).

Reply:
Understood. It's a Go-ism :)